File: ts/engine/test/normalize.test.ts
import { describe, expect, it } from 'vitest';
import {
  canonicalDate,
  canonicalInstant,
  canonicalTimestamp,
  normalizeRow,
  normalizeValue,
  type ColumnKinds,
} from '../src/normalize';

/**
 * Unit-level guards for the canonical temporal forms.
 *
 * The corpus is the contract and it runs against real databases; this file pins the two decisions a
 * corpus case cannot reach, because both are about what happens when the engine does NOT know
 * something — and a golden can only assert what it does know.
 */
describe('canonical temporal forms', () => {
  // 2024-04-01T10:00:00 UTC. Its LOCAL components differ from its UTC components on any machine that
  // is not on UTC, which is what makes the `components` argument observable.
  const value = new Date(Date.UTC(2024, 3, 1, 10, 0, 0));

  it('reads the half of the Date the driver put the digits in', () => {
    expect(canonicalTimestamp(value, 'utc')).toBe('2024-04-01T10:00:00');
    expect(canonicalDate(value, 'utc')).toBe('2024-04-01');
  });

  it('renders an instant in UTC, with the designator', () => {
    expect(canonicalInstant(value)).toBe('2024-04-01T10:00:00Z');
  });

  it('keeps a non-zero millisecond part and drops a zero one', () => {
    expect(canonicalTimestamp(new Date(Date.UTC(2024, 3, 1, 10, 0, 0, 250)), 'utc')).toBe(
      '2024-04-01T10:00:00.250',
    );
    expect(canonicalTimestamp(value, 'utc')).toBe('2024-04-01T10:00:00');
  });

  it('gives a DATE no invented midnight', () => {
    expect(canonicalDate(value, 'utc')).not.toContain('T');
  });
});

describe('normalizeValue', () => {
  const value = new Date(Date.UTC(2024, 3, 1, 10, 0, 0));

  /**
   * THE REFUSAL. An unidentified temporal is returned untouched rather than guessed at, because
   * guessing is what produced the bug this module was rewritten to fix: every `Date` used to be
   * flattened into the zone-less form, which silently shifted a `timestamptz` into the reader's zone
   * and gave a `DATE` a midnight it never had.
   */
  it('leaves a Date alone when the column kind is unknown', () => {
    expect(normalizeValue(value)).toBe(value);
    expect(normalizeValue(value, undefined)).toBe(value);
  });

  it('never touches a value that is not a Date', () => {
    // Including a string that LOOKS like a timestamp — a `note` column holding "2024-04-01" is text
    // a user typed, and rewriting it to look tidy would be data corruption.
    for (const other of ['2024-04-01', '19.99', 42, 0n, true, null, undefined]) {
      expect(normalizeValue(other, 'naive')).toBe(other);
    }
  });
});

describe('normalizeRow', () => {
  it('normalizes only the columns it was given kinds for, and keeps key order', () => {
    const kinds: ColumnKinds = new Map([
      ['created', 'naive'],
      ['due', 'date'],
      ['seen_at', 'instant'],
    ]);
    const at = new Date(Date.UTC(2024, 3, 1, 10, 0, 0));
    const row = { id: 7, created: at, due: at, seen_at: at, untyped: at, note: 'hello' };

    const out = normalizeRow(row, kinds, 'utc') as Record<string, unknown>;

    expect(Object.keys(out)).toEqual(['id', 'created', 'due', 'seen_at', 'untyped', 'note']);
    expect(out.created).toBe('2024-04-01T10:00:00');
    expect(out.due).toBe('2024-04-01');
    expect(out.seen_at).toBe('2024-04-01T10:00:00Z');
    // No kind was supplied for `untyped`, so it survives as the driver's own object.
    expect(out.untyped).toBe(at);
    expect(out.note).toBe('hello');
  });
});

File: dart/engine/test/normalize_test.dart
@TestOn('vm')
library;

import 'package:sqleasy_engine/src/normalize.dart';
import 'package:test/test.dart';

/// Pure form tests — no database. The DB-backed proof that each executor supplies the RIGHT kind
/// lives in the `integration`-tagged corpus replay; this pins the forms themselves, and the two
/// decisions the corpus cannot reach: an unknown kind is left alone, and an instant column whose
/// driver did not hand over an instant is left alone too.
///
/// Every case is stated twice where it matters — once UTC-flagged, once local-flagged — because the
/// whole point of reading a `DateTime`'s OWN components is that the two agree.
void main() {
  group('canonicalTimestamp — zone-less, no designator', () {
    test('writes the stored digits from either half, and never a Z', () {
      expect(canonicalTimestamp(DateTime.utc(2024, 4, 1, 10)),
          '2024-04-01T10:00:00');
      expect(
          canonicalTimestamp(DateTime(2024, 4, 1, 10)), '2024-04-01T10:00:00');
    });

    test('writes a whole second plainly and keeps real sub-second digits', () {
      expect(canonicalTimestamp(DateTime.utc(2024, 4, 1, 10, 0, 0, 0, 0)),
          '2024-04-01T10:00:00');
      expect(canonicalTimestamp(DateTime.utc(2024, 4, 1, 10, 0, 0, 500)),
          '2024-04-01T10:00:00.500');
      expect(canonicalTimestamp(DateTime.utc(2024, 4, 1, 10, 0, 0, 123, 456)),
          '2024-04-01T10:00:00.123456');
    });
  });

  group('canonicalDate — a date, with no invented midnight', () {
    test('is YYYY-MM-DD and nothing more', () {
      expect(canonicalDate(DateTime.utc(2024, 4, 1)), '2024-04-01');
      expect(canonicalDate(DateTime(2024, 4, 1)), '2024-04-01');
    });

    test('does not slide a day when the value is a UTC-flagged midnight', () {
      // `package:postgres` returns DATE '2024-04-01' exactly like this. Reading the local half
      // instead lands on 2024-03-31 anywhere west of Greenwich.
      expect(canonicalDate(DateTime.utc(2024, 4, 1)), '2024-04-01');
    });
  });

  group('canonicalInstant — UTC with the Z designator', () {
    test('keeps the instant and states it in UTC', () {
      expect(canonicalInstant(DateTime.utc(2024, 4, 1, 10)),
          '2024-04-01T10:00:00Z');
      expect(canonicalInstant(DateTime.utc(2024, 4, 1, 14)),
          '2024-04-01T14:00:00Z');
    });
  });

  group('normalizeValue', () {
    test('rewrites each kind into the form that kind actually means', () {
      final value = DateTime.utc(2024, 4, 1, 10);
      expect(
          normalizeValue(value, TemporalKind.instant), '2024-04-01T10:00:00Z');
      expect(normalizeValue(value, TemporalKind.naive), '2024-04-01T10:00:00');
      expect(normalizeValue(value, TemporalKind.date), '2024-04-01');
    });

    test('leaves a DateTime of UNKNOWN kind exactly as the driver made it', () {
      final value = DateTime.utc(2024, 4, 1, 10);
      expect(normalizeValue(value), same(value));
    });

    test('leaves a local-flagged value on an instant column alone', () {
      // A BACKSTOP, not a live path: no executor mapping produces this today. A driver that stamps
      // an instant column's value with the READER's zone has not handed over an instant — reading it
      // as one would give an answer that changes with TZ — so the driver's own object is reported
      // instead. This is the shape the retired MySQL `TIMESTAMP` -> instant mapping had.
      final value = DateTime(2024, 4, 1, 10);
      expect(normalizeValue(value, TemporalKind.instant), same(value));
    });

    test('passes non-temporal values through untouched', () {
      expect(normalizeValue(null), isNull);
      expect(normalizeValue(9007199254740993), 9007199254740993);
      expect(normalizeValue('2024-04-01'), '2024-04-01');
      expect(normalizeValue(true), true);
    });
  });

  group('normalizeRow / normalizeRows', () {
    test('keys and their order survive', () {
      final row = normalizeRow(
        {'placed_at': DateTime.utc(2024, 4, 1, 10), 'note': 'first order'},
        {'placed_at': TemporalKind.naive},
      );
      expect(row.keys.toList(), ['placed_at', 'note']);
      expect(row['placed_at'], '2024-04-01T10:00:00');
    });

    test('a column absent from the kinds map is not rewritten', () {
      final value = DateTime.utc(2024, 4, 1, 10);
      final row = normalizeRow({'when': value}, {'other': TemporalKind.naive});
      expect(row['when'], same(value));
    });

    test('with no kinds at all, nothing temporal moves', () {
      final value = DateTime.utc(2024, 4, 1, 10);
      expect(
          normalizeRows([
            {'a': value}
          ]),
          [
            {'a': value}
          ]);
    });
  });
}
